File: src/main/java/mods/eln/sim/mna/component/VoltageSource.java
package mods.eln.sim.mna.component;

import mods.eln.misc.INBTTReady;
import mods.eln.sim.mna.SubSystem;
import mods.eln.sim.mna.misc.ISubSystemProcessI;
import mods.eln.sim.mna.state.CurrentState;
import mods.eln.sim.mna.state.State;
import net.minecraft.nbt.NBTTagCompound;


public class VoltageSource extends Bipole implements ISubSystemProcessI, INBTTReady {

    String name;

    double voltage = 0;
    private final CurrentState currentState = new CurrentState();

    public VoltageSource(String name) {
        this.name = name;
    }

    public VoltageSource(String name, State aPin, State bPin) {
        super(aPin, bPin);
        this.name = name;
    }

    public VoltageSource setVoltage(double voltage) {
        this.voltage = voltage;
        return this;
    }

    public double getVoltage() {
        return voltage;
    }

    @Override
    public void quitSubSystem() {
        subSystem.states.remove(getCurrentState());
        subSystem.removeProcess(this);
        super.quitSubSystem();
    }

    @Override
    public void addToSubsystem(SubSystem s) {
        super.addToSubsystem(s);
        s.addState(getCurrentState());
        s.addProcess(this);
    }

    @Override
    public void applyToSubsystem(SubSystem s) {
        s.addToA(aPin, getCurrentState(), 1.0);
        s.addToA(bPin, getCurrentState(), -1.0);
        s.addToA(getCurrentState(), aPin, 1.0);
        s.addToA(getCurrentState(), bPin, -1.0);
    }

    @Override
    public void simProcessI(SubSystem s) {
        s.addToI(getCurrentState(), voltage);
    }

    @Override
    public double getCurrent() {
        return -getCurrentState().state;
    }

    public CurrentState getCurrentState() {
        return currentState;
    }

    @Override
    public void readFromNBT(NBTTagCompound nbt, String str) {
        str += name;
        setVoltage(nbt.getDouble(str + "U"));
        currentState.state = (nbt.getDouble(str + "Istate"));
    }

    @Override
    public void writeToNBT(NBTTagCompound nbt, String str) {
        str += name;
        nbt.setDouble(str + "U", voltage);
        nbt.setDouble(str + "Istate", currentState.state);
    }

    public double getPower() {
        return getVoltage() * getCurrent();
    }
}

File: src/main/java/mods/eln/sim/mna/component/VoltageSource.kt
package mods.eln.sim.mna.component

import mods.eln.misc.INBTTReady
import mods.eln.sim.mna.SubSystem
import mods.eln.sim.mna.misc.ISubSystemProcessI
import mods.eln.sim.mna.state.CurrentState
import mods.eln.sim.mna.state.State
import net.minecraft.nbt.NBTTagCompound


open class VoltageSource : Bipole, ISubSystemProcessI, INBTTReady {

    val currentState = CurrentState()

    var u: Double = 0.0

    override fun getCurrent(): Double {
        return -currentState.state
    }

    open fun getPower(): Double {
        return getVoltage() * getCurrent()
    }

    constructor(name: String) {
        this.name = name
    }

    constructor(name: String, aPin: State?, bPin: State?) : super(aPin, bPin) {
        this.name = name
    }

    override fun quitSubSystem() {
        getSubSystem()!!.states.remove(currentState)
        getSubSystem()!!.removeProcess(this)
        super.quitSubSystem()
    }

    override fun addedTo(s: SubSystem) {
        super.addedTo(s)
        s.addState(currentState)
        s.addProcess(this)
    }

    override fun applyTo(s: SubSystem) {
        s.addToA(aPin, currentState, 1.0)
        s.addToA(bPin, currentState, -1.0)
        s.addToA(currentState, aPin, 1.0)
        s.addToA(currentState, bPin, -1.0)
    }

    override fun simProcessI(s: SubSystem) {
        s.addToI(currentState, u)
    }

    // TODO: Move to a VoltageSourceNbt abstraction class
    override fun readFromNBT(nbt: NBTTagCompound, str: String) {
        u = (nbt.getDouble(str + name + "U"))
        currentState.state = nbt.getDouble(str + name + "Istate")
    }

    override fun writeToNBT(nbt: NBTTagCompound, str: String) {
        nbt.setDouble(str + name + "U", getVoltage())
        nbt.setDouble(str + name + "Istate", currentState.state)
    }
}
